<!-- -*- mode: XML -*- -->
<!-- run velodyne_pointcloud/CloudNodelet in a nodelet manager for an VLP-16 -->

<launch>
    <group>
        <push-ros-namespace namespace="middle_left"/>
        <!-- declare arguments with default values -->
        <arg name="launch_driver" default="true"/>

        <arg name="calibration" default="$(find-pkg-share velodyne_pointcloud)/params/VLP16db.yaml"/>
        <arg name="device_ip" default=""/>
        <arg name="frame_id" default="velodyne"/>
        <arg name="manager" default="$(var frame_id)_nodelet_manager"/>
        <arg name="max_range" default="130.0"/>
        <arg name="min_range" default="0.4"/>
        <arg name="pcap" default=""/>
        <arg name="port" default="2370"/>
        <arg name="read_fast" default="false"/>
        <arg name="read_once" default="false"/>
        <arg name="repeat_delay" default="0.0"/>
        <arg name="rpm" default="600.0"/>
        <arg name="scan_phase" default="0.0"/>
        <arg name="scan_period" default="100"/> <!-- ms -->
        <arg name="sensor_timestamp" default="false"/>
        <arg name="laserscan_ring" default="-1"/>
        <arg name="laserscan_resolution" default="0.007"/>
        <arg name="num_points_threshold" default="300"/>
        <arg name="invalid_intensity"
             default="[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]"/>
        <arg name="invalid_point_remove" default="true"/>
        <arg name="invalid_rings" default="[0,1,2,3,4]"/>
        <arg name="invalid_angles_start" default="[28000,28000,28000,3330,3330]"/>
        <arg name="invalid_angles_end" default="[29875,29875,29875,3635,3635]"/>

        <!-- start nodelet manager -->
        <include file="$(find-pkg-share velodyne_driver)/launch/nodelet_manager.launch.xml" if="$(var launch_driver)">
            <arg name="device_ip" value="$(var device_ip)"/>
            <arg name="frame_id" value="$(var frame_id)"/>
            <arg name="manager" value="$(var manager)"/>
            <arg name="model" value="VLP16"/>
            <arg name="pcap" value="$(var pcap)"/>
            <arg name="port" value="$(var port)"/>
            <arg name="read_fast" value="$(var read_fast)"/>
            <arg name="read_once" value="$(var read_once)"/>
            <arg name="repeat_delay" value="$(var repeat_delay)"/>
            <arg name="rpm" value="$(var rpm)"/>
            <arg name="scan_phase" value="$(var scan_phase)"/>
            <arg name="sensor_timestamp" value="$(var sensor_timestamp)"/>
        </include>

        <!-- start cloud nodelet -->
        <include file="$(find-pkg-share velodyne_pointcloud)/launch/cloud_nodelet.launch.xml">
            <arg name="calibration" value="$(var calibration)"/>
            <arg name="manager" value="$(var manager)"/>
            <arg name="max_range" value="$(var max_range)"/>
            <arg name="min_range" value="$(var min_range)"/>
            <arg name="num_points_threshold" value="$(var num_points_threshold)"/>
            <arg name="invalid_intensity" value="$(var invalid_intensity)"/>
            <arg name="scan_phase" value="$(var scan_phase)"/>
            <arg name="scan_period" value="$(var scan_period)"/>
            <arg name="invalid_point_remove" value="$(var invalid_point_remove)"/>
            <arg name="invalid_rings" value="$(var invalid_rings)"/>
            <arg name="invalid_angles_start" value="$(var invalid_angles_start)"/>
            <arg name="invalid_angles_end" value="$(var invalid_angles_end)"/>
        </include>
    </group>
    <group>
        <push-ros-namespace namespace="middle_right"/>
        <!-- declare arguments with default values -->
        <arg name="launch_driver" default="true"/>

        <arg name="calibration" default="$(find-pkg-share velodyne_pointcloud)/params/VLP16db.yaml"/>
        <arg name="device_ip" default=""/>
        <arg name="frame_id" default="velodyne"/>
        <arg name="manager" default="$(var frame_id)_nodelet_manager"/>
        <arg name="max_range" default="130.0"/>
        <arg name="min_range" default="0.4"/>
        <arg name="pcap" default=""/>
        <arg name="port" default="2372"/>
        <arg name="read_fast" default="false"/>
        <arg name="read_once" default="false"/>
        <arg name="repeat_delay" default="0.0"/>
        <arg name="rpm" default="600.0"/>
        <arg name="scan_phase" default="0.0"/>
        <arg name="scan_period" default="100"/> <!-- ms -->
        <arg name="sensor_timestamp" default="false"/>
        <arg name="laserscan_ring" default="-1"/>
        <arg name="laserscan_resolution" default="0.007"/>
        <arg name="num_points_threshold" default="300"/>
        <arg name="invalid_intensity"
             default="[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]"/>
        <arg name="invalid_point_remove" default="true"/>
        <arg name="invalid_rings" default="[0,1,2]"/>
        <arg name="invalid_angles_start" default="[0]"/>
        <arg name="invalid_angles_end" default="[0]"/>

        <!-- start nodelet manager -->
        <include file="$(find-pkg-share velodyne_driver)/launch/nodelet_manager.launch.xml" if="$(var launch_driver)">
            <arg name="device_ip" value="$(var device_ip)"/>
            <arg name="frame_id" value="$(var frame_id)"/>
            <arg name="manager" value="$(var manager)"/>
            <arg name="model" value="VLP16"/>
            <arg name="pcap" value="$(var pcap)"/>
            <arg name="port" value="$(var port)"/>
            <arg name="read_fast" value="$(var read_fast)"/>
            <arg name="read_once" value="$(var read_once)"/>
            <arg name="repeat_delay" value="$(var repeat_delay)"/>
            <arg name="rpm" value="$(var rpm)"/>
            <arg name="scan_phase" value="$(var scan_phase)"/>
            <arg name="sensor_timestamp" value="$(var sensor_timestamp)"/>
        </include>

        <!-- start cloud nodelet -->
        <include file="$(find-pkg-share velodyne_pointcloud)/launch/cloud_nodelet.launch.xml">
            <arg name="calibration" value="$(var calibration)"/>
            <arg name="manager" value="$(var manager)"/>
            <arg name="max_range" value="$(var max_range)"/>
            <arg name="min_range" value="$(var min_range)"/>
            <arg name="num_points_threshold" value="$(var num_points_threshold)"/>
            <arg name="invalid_intensity" value="$(var invalid_intensity)"/>
            <arg name="scan_phase" value="$(var scan_phase)"/>
            <arg name="scan_period" value="$(var scan_period)"/>
            <arg name="invalid_point_remove" value="$(var invalid_point_remove)"/>
            <arg name="invalid_rings" value="$(var invalid_rings)"/>
            <arg name="invalid_angles_start" value="$(var invalid_angles_start)"/>
            <arg name="invalid_angles_end" value="$(var invalid_angles_end)"/>
        </include>
    </group>
</launch>
